pipelines/tekton/pipeplines_provider.go

```diff
@@ -0,0 +1,199 @@
+package tekton
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/tektoncd/cli/pkg/pipelinerun"
+	"github.com/tektoncd/cli/pkg/taskrun"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	fn "knative.dev/kn-plugin-func"
+	"knative.dev/kn-plugin-func/k8s"
+	"knative.dev/kn-plugin-func/knative"
+	"knative.dev/kn-plugin-func/pipelines"
+	"knative.dev/pkg/apis"
+)
+
+type Opt func(*PipelinesProvider) error
+
+type PipelinesProvider struct {
+	// namespace with which to override that set on the default configuration (such as the ~/.kube/config).
+	// If left blank, pipeline creation/run will commence to the configured namespace.
+	namespace                            string
+	Verbose                              bool
+	progressListener                     fn.ProgressListener
+	containerRegistryCredentialsCallback pipelines.ContainerRegistryCredentialsCallback
+}
+
+func WithNamespace(namespace string) Opt {
+	return func(pp *PipelinesProvider) error {
+		namespace, err := k8s.GetNamespace(namespace)
+		if err != nil {
+			return err
+		}
+		pp.namespace = namespace
+		return nil
+	}
+}
+
+func WithProgressListener(pl fn.ProgressListener) Opt {
+	return func(pp *PipelinesProvider) error {
+		pp.progressListener = pl
+		return nil
+	}
+}
+
+func WithPromptForContainerRegistryCredentials(cbk pipelines.ContainerRegistryCredentialsCallback) Opt {
+	return func(pp *PipelinesProvider) error {
+		pp.containerRegistryCredentialsCallback = cbk
+		return nil
+	}
+}
+
+func NewPipelinesProvider(opts ...Opt) (*PipelinesProvider, error) {
+	pp := &PipelinesProvider{}
+
+	for _, opt := range opts {
+		err := opt(pp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pp, nil
+}
+
+func (pp *PipelinesProvider) Run(ctx context.Context, f fn.Function) error {
+	pp.progressListener.Increment("Creating Pipeline resources")
+
+	client, err := NewTektonClient()
+	if err != nil {
+		return err
+	}
+
+	err = k8s.CreatePersistentVolumeClaim(ctx, getPipelinePvcName(f), pp.namespace, corev1.ReadWriteOnce, *resource.NewQuantity(500, resource.DecimalSI))
```

Comment:
Is 500 bytes enough (I don't what is being stored there)?

Reply:
Isn't this where the function code is cloned?

Reply:
Yeah, it holds the source code and cache. It means `500 MiB` not `500 bytes` though :)

Reply:
That weird.
```go
func TestPipelinesProvider_Run(t *testing.T) {
	var q *resource.Quantity
	q = resource.NewQuantity(500*(1_000_000), resource.DecimalSI)
	t.Logf("%+v", q)
	q = resource.NewQuantity(500, resource.DecimalSI)
	t.Logf("%+v", q)
}
```
Output:
```
=== RUN   TestPipelinesProvider_Run
    pipeplines_provider_test.go:11: 500M
    pipeplines_provider_test.go:13: 500
--- PASS: TestPipelinesProvider_Run (0.00s)
PASS
```
Maybe `k8s` multiplies the quantity by million? @zroubalik

Reply:
```go
	// Volume size, in bytes (e,g. 5Gi = 5GiB = 5 * 1024 * 1024 * 1024)
	ResourceStorage ResourceName = "storage"
```
really weird.

Reply:
```go
func TestPipelinesProvider_Run(t *testing.T) {

	pvc := &corev1.PersistentVolumeClaim{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "nm",
			Namespace: "nms",
		},
		Spec: corev1.PersistentVolumeClaimSpec{
			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
			Resources: corev1.ResourceRequirements{
				Requests: corev1.ResourceList{},
			},
		},
	}
	pvc.Spec.Resources.Requests[corev1.ResourceStorage] = *resource.NewQuantity(500, resource.DecimalSI)
	enc := json.NewEncoder(os.Stderr)
	enc.Encode(&pvc)

	pvc.Spec.Resources.Requests[corev1.ResourceStorage] = *resource.NewQuantity(500*(1<<20), resource.BinarySI)
	enc.Encode(&pvc)
}
```
Ouptut:
```
{"metadata":{"name":"nm","namespace":"nms","creationTimestamp":null},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"500"}}},"status":{}}
{"metadata":{"name":"nm","namespace":"nms","creationTimestamp":null},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"500Mi"}}},"status":{}}
```

Reply:
@matejvasek mabye it fallbacks to `Mi` if unit is not specified?

Reply:
@rhuss @dsimansk any idea how this works?

Reply:
I will explicitly specify the unit here, but the real question is whether 500Mi is ok value, it should store source code and buildpacks cache and layers.

Reply:
I have added here the default size (500Mi) for now:
https://github.com/knative-sandbox/kn-plugin-func/blob/22923c455ed46da35ede04ce19346edd652788c7/pipelines/tekton/defaults.go#L4-L6
Should be cofigurable in the future.